docs/guides/typescript.md

```diff
@@ -1,75 +1,213 @@
 ## Using TypeScript
 
-As XState is written in [TypeScript](https://www.typescriptlang.org/), strongly typing your statecharts is useful and encouraged. Consider this light machine example:
+As XState is written in [TypeScript](https://www.typescriptlang.org/), strongly typing your statecharts is useful and encouraged.
 
 ```typescript
-// The events that the machine handles
-type LightEvent =
-  | { type: 'TIMER' }
-  | { type: 'POWER_OUTAGE' }
-  | { type: 'PED_COUNTDOWN'; duration: number };
-
-// The context (extended state) of the machine
-interface LightContext {
-  elapsed: number;
-}
+import { t, createMachine } from 'xstate';
+
+const lightMachine = createMachine({
+  schema: {
+    // The context (extended state) of the machine
+    context: t<{ elapsed: number }>(),
+    // The events this machine handles
+    events: t<
+      | { type: 'TIMER' }
+      | { type: 'POWER_OUTAGE' }
+      | { type: 'PED_COUNTDOWN'; duration: number }
+    >()
+  }
+  /* Other config... */
+});
+```
+
+Providing the context and events to the `schema` attribute gives many advantages:
+
+- The context type/interface (`TContext`) is passed on to actions, guards, services and more. It is also passed to deeply nested states.
+- The event type (`TEvent`) ensures that only specified events (and built-in XState-specific ones) are used in transition configs. The provided event object shapes are also passed on to actions, guards, and services.
+- Events which you send to the machine will be strongly typed, offering you much more confidence in the payload shapes you'll be receiving.
+
+## Typegen with the VSCode Extension <Badge text="4.29+" />
 
-const lightMachine = createMachine<LightContext, LightEvent>({
-  key: 'light',
-  initial: 'green',
-  context: { elapsed: 0 },
+Using our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode), you can automatically generate intelligent typings for XState.
+
+> This feature is in beta! See the section on known limitations below to see what we're actively looking to improve.
```

Comment:
```suggestion
::: warning
This feature is in beta! See the section on known limitations below to see what we're actively looking to improve.
```

Reply:
Changed it to this:

![image](https://user-images.githubusercontent.com/28293365/151373174-0bd06775-111e-4055-a77d-de1057121bfa.png)